tensorboard/data/provider.py

```diff
@@ -0,0 +1,223 @@
+
+    Args:
+      experiment_id: ID of enclosing experiment.
+      owner_plugin: String name of the TensorBoard plugin that created
+        the data to be queried. Required.
+      run_tag_filter: Optional `RunTagFilter` value. If omitted, all
+        runs and tags will be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a `ScalarMetadata`
+      value.
+    """
+    pass
+
+  @abc.abstractmethod
+  def read_scalars(
+      self,
+      experiment_id,
+      owner_plugin,
+      downsample_to=None,
+      run_tag_filter=None,
+      step_filter=None,
+  ):
+    """Read values from scalar time series.
+
+    Args:
+      experiment_id: ID of enclosing experiment.
+      owner_plugin: String name of the TensorBoard plugin that created
+        the data to be queried. Required.
+      downsample_to: Integer number of steps to which to downsample the
+        results (e.g., `1000`). Required.
+      run_tag_filter: Optional `RunTagFilter` value. If provided, a time
+        series will only be included in the result if its run and tag
+        both pass this filter. If `None`, all time series will be
+        included.
+      step_filter: Optional `StepFilter` value. If `None`, the entire
+        range of steps may be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a list of
+      `ScalarDatum` values sorted by step.
+    """
+    pass
+
+  def list_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def read_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def list_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+  def read_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+
+_ScalarMetadata = collections.namedtuple(
+    "ScalarMetadata",
+    ("max_step", "corresponding_wall_time", "summary_metadata"),
+)
+class ScalarMetadata(_ScalarMetadata):
+  """Metadata about a scalar time series for a particular run and tag.
+
+  Attributes:
+    max_step: The largest step for this scalar time series; an integer.
+    corresponding_wall_time: The wall time of the datum with the largest
+      step in this time series, as `float` seconds since epoch.
+    summary_metadata: A `summary_pb2.SummaryMetadata` value.
+  """
+  pass
+
+
+_ScalarDatum = collections.namedtuple(
+    "ScalarDatum", ("step", "wall_time", "value")
+)
+class ScalarDatum(_ScalarDatum):
+  """A single datum in a scalar time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This
+      is a unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as
```

Comment:
Since we have the opportunity to newly define this field, do we want to stick with float, vs datetime or integer microseconds?  No particularly strong preference on my part but I thought it was worth asking.

Reply:
I’m somewhat disinclined to use `datetime.datetime` because of the whole
“naïve vs. aware” distinction with poor time zone support in the
standard library, and issues like #2017. See #2054, where we switched
_from_ `datetime` to bare timestamps.

Timestamps represented as “f64 seconds” have full `1e-6` precision
through year 2100, so the raw precision isn’t a big concern. I could be
convinced to use integer microseconds, but afaik that’d be a new pattern
in TensorBoard, so I’m weakly inclined to keep this as is.

Reply:
That's a reasonable argument to stick with double/float64, sounds good to me.